File: Data Structures & Algorithms/daily-temperatures/submission-4.cpp
class Solution {
public:
    vector<int> dailyTemperatures(vector<int>& temperatures) {
        vector<int> res(temperatures.size());
        stack<pair<int, int>> tracker;
        for(int i = temperatures.size() - 1; i >= 0; i--){
            while(!tracker.empty() && tracker.top().first <= temperatures[i]){
                tracker.pop();
            }

            if(!tracker.empty() && temperatures[i] < tracker.top().first){
                res[i] = tracker.top().second - i;
            }
            tracker.push({temperatures[i], i});
        }
        return res;
    }
};

File: Data Structures & Algorithms/sliding-window-maximum/submission-4.py
class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        res = []
        tracker = deque([])
        i = 0
        for j in range(len(nums)):
            while len(tracker) > 0 and tracker[-1][0] < nums[j]:
                tracker.pop()
            tracker.append((nums[j], j))
            while len(tracker) > 0 and tracker[0][1] < i:
                tracker.popleft()
            if j + 1 >= k:
                res.append(tracker[0][0])
                i += 1
        return res